### DISCUSSION TITLE
Automatically regenerate media sizes

### DISCUSSION THREAD

**Discussion by jmikrut** [payloadcms/payload#1834] [Feature Requests & Ideas]
This is a commonly asked topic and it just came up in Discord today, so I figured I would share a quick script that the team and I use for our Payload projects when we need something like WP's "Regenerate Thumbnails" plugin. Basically if you update your `imageSizes` and want to automatically re-run the image transformations for each one of your uploads, this can be super handy.

```js
const payload = require('payload');
const path = require('path');

require('dotenv').config();

const { PAYLOAD_SECRET_KEY, MONGO_URL } = process.env;

payload.init({
  secret: PAYLOAD_SECRET_KEY,
  mongoURL: MONGO_URL,
  local: true,
});

const regenerateMediaSizes = async () => {
  try {
    const media = await payload.find({
      collection: 'media',
      depth: 0,
      limit: 300,
    });
    await Promise.all(media.docs.map(async (mediaDoc) => {
      try {
        const staticDir = path.resolve(__dirname, '../media');

        await payload.update({
          collection: 'media',
          id: mediaDoc.id,
          data: mediaDoc,
          filePath: `${staticDir}/${mediaDoc.filename}`,
          overwriteExistingFiles: true,
        });

        console.log(`Media ${mediaDoc.alt || mediaDoc.id} regenerated successfully`);
      } catch (err) {
        console.error(`Media ${mediaDoc.alt || mediaDoc.id} failed to regenerate`);
        console.error(err);
      }
    }));
  } catch (err) {
    console.log('Unable to find documents with payload');
    console.error(err);
    process.exit(0);
  }

  console.log('Media size regeneration completed!');
  process.exit(0);
};

regenerateMediaSizes();
```

If you save this file as `regenerateMediaSizes.js`, then all you need to do to run it is `node ./regenerateMediaSizes.js` and boom.

Related: #434

**Comment by fabioquarantini** [comment]
How to delete old unused thumbnails?

**Comment by ChristopherNowlan** [comment]
I had to move payload init function into the regenerateMediaSizes async function. Because my config file was within the src directory I had to update the command to be `cross-env PAYLOAD_CONFIG_PATH=dist/payload.config.js node ./regenerateMediaSizes.js`


```
const payload = require('payload');
const path = require('path');

require('dotenv').config();

const { PAYLOAD_SECRET_KEY, MONGO_URL } = process.env;

const regenerateMediaSizes = async () => {
  try {

  await payload.init({
        secret: PAYLOAD_SECRET_KEY,
        mongoURL: MONGODB_URL,
        local: true,
    });

    const media = await payload.find({
      collection: 'media',
      depth: 0,
      limit: 300,
    });

    await Promise.all(media.docs.map(async (mediaDoc) => {
      try {
        const staticDir = path.resolve(__dirname, '../media');

        await payload.update({
          collection: 'media',
          id: mediaDoc.id,
          data: mediaDoc,
          filePath: `${staticDir}/${mediaDoc.filename}`,
          overwriteExistingFiles: true,
        });

        console.log(`Media ${mediaDoc.alt || mediaDoc.id} regenerated successfully`);
      } catch (err) {
        console.error(`Media ${mediaDoc.alt || mediaDoc.id} failed to regenerate`);
        console.error(err);
      }
    }));
  } catch (err) {
    console.log('Unable to find documents with payload');
    console.error(err);
    process.exit(0);
  }

  console.log('Media size regeneration completed!');
  process.exit(0);
};

regenerateMediaSizes();
```

**Comment by andriasang** [comment]
@jmikrut : Is this script compatible with the [cloud storage plugin](https://github.com/payloadcms/plugin-cloud-storage)?  I removed the "filePath" section of the update call and it seems that every image in my collection is being iterated over but the sizes area is just  empty objects:

 sizes: { w400: {}, w768: {}, w1024: {} },

**Comment by arieltonglet** [comment]
@andriasang I've managed to make a version of this script that works with [plugin-cloud-storage](https://github.com/payloadcms/plugin-cloud-storage/blob/master/docs/local-dev.md) – building on top of @ChristopherNowlan script and with a huge help from @jmikrut [on Discord](https://discord.com/channels/967097582721572934/1080843016413118544/1080843016413118544).

**The main point to get it working is**: get the media `url`, fetch it, and pass it as a [File](https://github.com/payloadcms/payload/blob/4d578f1bfd05efab5cc8db95895eabb776b2d9d1/src/uploads/types.ts#L82) in a `file` prop instead of `filePath`

**Important** to note as well: Check if you have the latest config file in your `dist` folder, to get the updated imageSizes. To make sure of that, I added a `build` to the script. Its more of a local development problem than a production one, but just in case:

`yarn build && cross-env PAYLOAD_CONFIG_PATH=dist/payload.config.js node ./regenerateMediaSizes.js`

```
const payload = require("payload");
const path = require("path");

require("dotenv").config();

const { PAYLOAD_SECRET, MONGODB_URI } = process.env;

const regenerateMediaSizes = async () => {
  try {
    await payload.init({
      secret: PAYLOAD_SECRET,
      mongoURL: MONGODB_URI,
      local: true,
    });

    const media = await payload.find({
      collection: "media",
      depth: 0,
      limit: 300,
    });

    await Promise.all(
      media.docs.map(async (mediaDoc) => {
        try {
          await fetch(mediaDoc.url)
            .then((response) => response.blob())
            .then(async (blob) => {
              const arrayBuffer = await blob.arrayBuffer();
              const buffer = await Buffer.from(arrayBuffer, "binary");

              const file = {
                data: buffer,
                mimetype: blob.type,
                name: mediaDoc.filename,
                size: mediaDoc.filesize,
              };

              await payload.update({
                collection: "media",
                id: mediaDoc.id,
                data: mediaDoc,
                file,
                overwriteExistingFiles: true,
                contentType: blob.type,
              });

              console.log(
                `Media ${mediaDoc.id} (${mediaDoc.filename}) regenerated successfully`
              );
            });
        } catch (err) {
          console.error(
            `Media ${mediaDoc.id} (${mediaDoc.filename}) failed to regenerate`
          );
          console.error(err);
        }
      })
    );
  } catch (err) {
    console.log("Unable to find documents with payload");
    console.error(err);
    process.exit(0);
  }

  console.log("Media size regeneration completed!");
  process.exit(0);
};

regenerateMediaSizes();
```

**Comment by linobino1** [comment]
When I use @jmikrut's script*, images get lost in a random way. Often it works as expected, but sometimes the original file is removed during the process.

\* adapted to work with payload 1.6.27 async payload.init()

**Reply by richadr** [reply]
try with `overwriteExistingFiles: false`

**Comment by johnrisby** [comment]
Hi, is there a problem running this under the current Payload (I'm on 1.15.7) ? I'm getting errors about not finding the collection. I've spent a while trying various things but still not working. If it does work, can someone explain where it should be run from etc please. Thanks

**Reply by isamardzija** [reply]
Same issue here. Not sure where I'm supposed to run this file from.

**Reply by ChristopherNowlan** [reply]
@johnrisby Can I see what your file structure looks. It might help me resolve your issue.  I just set it up as a npm script or you could hook it into the build script like @andriasang

**Reply by ChristopherNowlan** [reply]
When I did this I had media and regenerateMediaSizes.js file up a level outside of the src directory

**Reply by danpii** [reply]
Hi, I am having the same issue. 
```
Unable to find documents with payload
APIError: The collection with slug media can't be found. Find Operation.
    at findLocal (/path/to/folder/payload-test/payload-0/node_modules/payload/dist/collections/operations/local/find.js:23:15)
    at Payload.find (/path/to/folder/payload-test/payload-0/node_modules/payload/dist/payload.js:70:16)
    at regenerateMediaSizes (/path/to/folder/payload-test/payload-0/regenerateMediaSizes.js:24:37)
    at Object.<anonymous> (/path/to/folder/payload-test/payload-0/regenerateMediaSizes.js:58:1)
```
This is my structure:
<img width="351" alt="image" src="https://github.com/payloadcms/payload/assets/16837118/7cfa9802-14af-469b-9b94-e226b41bcb49">


edit: sorry, removed sensible data and posted again

**Reply by danpii** [reply]
> When I did this I had media and regenerateMediaSizes.js file up a level outside of the src directory

I have tried it with regenerateMediaSizes.js outside of the src directory as well. Same error unfortunately

**Reply by ChristopherNowlan** [reply]
And the media directory as well?  Does you regenerate file look like below? Version of payload are you using? 

`const payload = require('payload');
const path = require('path');

require('dotenv').config();

const { PAYLOAD_SECRET, MONGODB_URI } = process.env;

const regenerateMediaSizes = async () => {

    try {

        await payload.init({
            secret: PAYLOAD_SECRET,
            mongoURL: MONGODB_URI,
            local: true,
        });

        const media = await payload.find({
            collection: 'media',
            depth: 0,
            limit: 300,
        })

        await Promise.all(media.docs.map(async (mediaDoc) => {
            try {
                const staticDir = path.resolve(__dirname, './media');

                await payload.update({
                    collection: 'media',
                    id: mediaDoc.id,
                    data: mediaDoc,
                    filePath: `${staticDir}/${mediaDoc.filename}`,
                    overwriteExistingFiles: true,
                });

                console.log(`Media ${mediaDoc.alt || mediaDoc.id} regenerated successfully`);
            } catch (err) {
                console.error(`Media ${mediaDoc.alt || mediaDoc.id} failed to regenerate`);
                console.error(err);
            }
        }));
    } catch (err) {
        console.log('Unable to find documents with payload');
        console.error(err);
        process.exit(0);
    }

    console.log('Media size regeneration completed!');
    process.exit(0);
};

regenerateMediaSizes();
`

**Reply by danpii** [reply]
yes, just moved media directory as well.
when I move the payload init function into the regenerateMediaSizes async function, I get an error:
```
payload-0/src/payload.config.ts:1
import path from 'path'
^^^^^^

SyntaxError: Cannot use import statement outside a module

```
I am using TS not JS. I'll try some more tomorrow

Payload Version: 2.11.0

**Reply by ChristopherNowlan** [reply]
@danpii try making the regenerateMediaSizes file js

**Reply by danpii** [reply]
my regenerateMediaSizes is actually js. all the other files are ts.
the error points to payload.config.ts. And I thought that error is typically JS relevant.
I'll make my regenerateMediaSizes file ts and see how it goes.

**Reply by ChristopherNowlan** [reply]
Mine is js and everything else is ts. What version of payload are you using.

**Reply by danpii** [reply]
v2.11.0

**Reply by ChristopherNowlan** [reply]
@danpii are you working on locally? If you are, stop dev. Run yarn build, then start dev again, then run the script. I just tested on v2.11.0

**Reply by danpii** [reply]
@ChristopherNowlan thanks for your help, really appreciated! Yes, I am running locally. I tried your recommendation, but I still get the same error:
```
Unable to find documents with payload
/Users/danielpanagiotopulos/Documents/_Work/WebStorm/playgrounds/payload-test/payload-0/src/payload.config.ts:1
import path from 'path'
^^^^^^

SyntaxError: Cannot use import statement outside a module

```

**Reply by ChristopherNowlan** [reply]
No worries. Change to
/dist/payload.config.js

**Reply by ChristopherNowlan** [reply]
Because I got the same issue when I tried to run it through src/payload.config.ts

**Reply by danpii** [reply]
@ChristopherNowlan Awesome! That worked! I had always run the scipt directly, but now with a npm command and /dist/payload.config.js it's running! Was that the reason why you have 'media' at root level?

For future me, and maybe others (@johnrisby @isamardzija ) here a little how to, this is how it works for me now:
Tested with Payload v2.11.0 on localhost

- media folder and regenerateMediaSizes.js on payload root level (same where package.json lies)
- use regenerateMediaSizes.js from @ChristopherNowlan - see [his comment](https://github.com/payloadcms/payload/discussions/1834#discussioncomment-4989249)
- run yarn build
- add this npm command to "scripts" in package.json
    - `"regenerateMediaSizes": "cross-env PAYLOAD_CONFIG_PATH=dist/payload.config.js node ./regenerateMediaSizes.js"`
- then: `npm run regenerateMediaSizes`

info: when you change 'imageSizes' in your Media collection you have to `run yarn build` again

**Reply by ChristopherNowlan** [reply]
@danpii you could add the regenerate script to end of build script. That way it runs every time you build.

**Comment by 58bits** [comment]
Here's a version based on @arieltonglet 's  excellent script - also for  [plugin-cloud-storage](https://github.com/payloadcms/payload/tree/main/packages/plugin-cloud-storage ) .

In this version only one image is processed at a time. For concurrency (with rate limits) and a larger number of files we'd use https://github.com/sindresorhus/p-limit 

This is a Typescript file, that can be placed anywhere, but must be run via ts-node in the root of the Payload project (at the same level as the .env file)

For example... `ts-node src/scripts/regenerateMediaSizes.ts`


```ts
/* eslint-disable @typescript-eslint/ban-ts-comment */
// @ts-nocheck

import payload from 'payload'

// eslint-disable-next-line
require('dotenv').config()

const { PAYLOAD_SECRET, DATABASE_URI } = process.env

const regenerateMediaSizes = async (): Promise<void> => {
  await payload.init({
    secret: PAYLOAD_SECRET,
    mongoURL: DATABASE_URI,
    local: true
  })

  const media = await payload.find({
    collection: 'media',
    depth: 0,
    limit: 100
  })

  if (media != null && media.totalDocs > 0) {
    payload.logger.info(`Found ${media.totalDocs} media files.`)
    for (let index = 0; index < media.docs.length; index++) {
      const mediaDoc = media.docs[index]
      try {
        await fetch(mediaDoc.url)
          .then(async (response) => await response.blob())
          .then(async (blob) => {
            const arrayBuffer = await blob.arrayBuffer()
            const buffer = Buffer.from(arrayBuffer, 'binary')

            const file = {
              data: buffer,
              mimetype: blob.type,
              name: mediaDoc.filename,
              size: mediaDoc.filesize
            }

            await payload.update({
              collection: 'media',
              id: mediaDoc.id,
              data: mediaDoc,
              file,
              overwriteExistingFiles: true,
              contentType: blob.type
            })

            payload.logger.info(
              `Media ${mediaDoc.id} (${mediaDoc.filename}) regenerated successfully`
            )
          })
      } catch (err) {
        payload.logger.error(`Media ${mediaDoc.id} (${mediaDoc.filename}) failed to regenerate`)
        console.error(err)
      }
    }
  } else {
    payload.logger.info('No media files found.')
  }

  payload.logger.info('Done!')
  process.exit(0)
}

void regenerateMediaSizes()
```

**Comment by tsemachh** [comment]
For Payload 3 we used:
```typescript

import { getPayload } from 'payload'
import * as dotenv from 'dotenv'
import path from 'node:path'
import { fileURLToPath } from 'node:url'
import { importConfig } from 'payload/node'
dotenv.config() // Load environment variables from .env

const __dirname: string = path.dirname(fileURLToPath(import.meta.url))
const payloadConfigPath: string = path.join(__dirname, '../payload.config.ts')
const awaitedConfig = await importConfig(payloadConfigPath)
const payload = await getPayload({ config: awaitedConfig })

// eslint-disable-next-line
import 'dotenv/config'

const regenerateMediaSizes = async (): Promise<void> => {
  const media = await payload.find({
    collection: 'media',
    depth: 0,
    limit: 500,
  })
  if (media != null && media.totalDocs > 0) {
    payload.logger.info(`Found ${media.totalDocs} media files.`)
    for (let index = 0; index < media.docs.length; index++) {
      const mediaDoc = media.docs[index]
      const req = {
        headers: new Map(), // Or use any object literal
      } as PayloadRequestWithData
      req.headers.set('origin', process.env.PAYLOAD_URL)
      try {
        await payload.update({
          collection: 'media',
          id: mediaDoc.id,
          data: mediaDoc,
          overwriteExistingFiles: true,
          req,
        })

        payload.logger.info(
          `Media ${mediaDoc.id} (${mediaDoc.filename}) successfully and new copy created.`,
        )
      } catch (err) {
        payload.logger.error(`Media ${mediaDoc.id} (${mediaDoc.filename}) failed to regenerate`)
        console.error(err)
      }
    }
  } else {
    payload.logger.info('No media files found.')
  }

  payload.logger.info('Done!')
  process.exit(0)
}

void regenerateMediaSizes()

```

**Comment by alextreppass** [comment]
For Payload 2 I used something similar to the above approach - specifying `file` and overwrite existing

However, it fails to regenerate a certain image size we have 😭 
```js
      {
        name: 'hero',
        // ≈1.414:1
        width: 2048,
        height: 1448,
        position: 'centre',
      },
```

The other sizes were ok.

I ended up having to dump all the affected ids, and manually go to the media entry in our running app admin page, and, one by one:
- Hit 'Edit Image'
- Hit 'Apply changes' (even though no changes are made)
- Hit Save

The middle part of this seemed necessary else the sizes would not regenerate

**Comment by rvanderfeer** [comment]
I'm currently using this as a migration with Payload v3, version 3.33.0
Seems less than ideal but it does seem to do the trick.
I tried some of the solutions provided above but none worked for me as typings seem to have changed in the latest Payload v3.


```Typescript
import { MigrateDownArgs, MigrateUpArgs } from '@payloadcms/db-mongodb'
import { Payload, UploadCollectionSlug, UploadConfig } from 'payload'
import fs from 'fs'
import path from 'path'

async function regenerateImageSizes(payload: Payload, collectionSlug: UploadCollectionSlug) {
	// Get the collection config to access upload settings
	const collectionConfig = payload.collections[collectionSlug].config
	const uploadConfig = collectionConfig.upload as UploadConfig

	// Get all documents that have files
	const { docs } = await payload.find({
		collection: collectionSlug,
		depth: 0,
		limit: 1000, // Adjust as needed
	})

	console.log(`Found ${docs.length} documents to process`)

	// Update each document to trigger regeneration
	for (const doc of docs) {
		if (doc.filename) {
			// Only process documents with files
			console.log(`Processing document ${doc.id}...`)

			try {
				// Get the file path from PayloadCMS upload directory
				const { staticDir } = uploadConfig
				const filePath = path.join(staticDir!, doc.filename!)

				// Check if file exists
				if (!fs.existsSync(filePath)) {
					console.error(`File not found: ${filePath}`)
					continue
				}

				// Read the file as buffer
				const fileBuffer = fs.readFileSync(filePath)

				// Re-upload the file to trigger image processing
				await payload.update({
					collection: collectionSlug,
					id: doc.id,
					data: {
						alt: doc.alt?.replace(/\.[\w]+$/, ''),
					},
					file: {
						data: fileBuffer,
						size: fs.statSync(filePath).size,
						mimetype: doc.mimeType!,
						name: doc.filename,
					}
				})

				console.log(`Successfully regenerated sizes for document ${doc.id}`)
			} catch (error) {
				console.error(`Error processing document ${doc.id}:`, error)
			}
		}
	}

	console.log('Image regeneration complete!')
}

export async function up({ payload }: MigrateUpArgs): Promise<void> {
	// Collection containing your uploads
	const MEDIA_COLLECTIONS = ['media', 'brandbook'] as UploadCollectionSlug[]

	// Process each collection
	for (const collectionSlug of MEDIA_COLLECTIONS) {
		await regenerateImageSizes(payload, collectionSlug).catch(console.error)
	}
}

export async function down({}: MigrateDownArgs): Promise<void> {
	console.log('No down migration needed for image regeneration')
}

```

**Reply by camilo220** [reply]
Hi @rvanderfeer, I'm using the storageS3 plugin, and I'm trying to adapt a migration script to regenerate image sizes. Since the filepaths aren't local, I'm struggling with how to modify the script to work with S3. I tried a simple update with overwriteExistingFiles: true, but it's still not working. Do you have any ideas on how to make it function correctly?

**Reply by rvanderfeer** [reply]
I don't have any hands-on experience with PayloadCMS+S3 but I think you will need to use the aws-sdk package to retrieve the file, I think doc.filename should contain the 'key' to the file in the S3 bucket. So s3.getObject() should retrieve that and you should then be able to assign the s3 response's file buffer to the payload.update() method.

**Reply by camilo220** [reply]
Unfortunately, I'm still quite new to TypeScript and PayloadCMS, and I haven't managed to fully resolve this yet. However, I'll leave a draft migration script that successfully triggers image regeneration within PayloadCMS but does not upload the new files to S3. As a result, the images' new sizes are created and stored in MongoDB, but the files themselves are missing from the S3 bucket.

My current hypothesis is that I'm missing some configuration to enable the migration to properly utilize the Payload S3 storage plugin.

```
import { MigrateUpArgs, MigrateDownArgs } from '@payloadcms/db-mongodb';
import { S3Client, GetObjectCommand } from '@aws-sdk/client-s3';
import { UploadCollectionSlug } from 'payload';
import { getSignedUrl } from '@aws-sdk/s3-request-presigner';
import fetch from 'node-fetch';

export async function up({ payload }: MigrateUpArgs): Promise<void> {
  const s3 = new S3Client({
    region: process.env.S3_REGION,
    credentials: {
      accessKeyId: process.env.S3_ACCESS_KEY_ID!,
      secretAccessKey: process.env.S3_SECRET_ACCESS_KEY!,
    },
  });

  const collectionSlug: UploadCollectionSlug = 'blog-media';
  const { docs } = await payload.find({
    collection: collectionSlug,
    depth: 0,
    limit: 1000,
  });

  console.log(`🔍 Found ${docs.length} documents in '${collectionSlug}'`);

  for (const doc of docs) {
    if (!doc.filename) continue;

    const prefix = 'cms-data/blog-media';

    const originalFilename = doc.filename.replace(/-\d+x\d+(?=\.\w+$)/, '');
    const key = `${prefix}/${originalFilename}`;
    console.log(`Original file name: ${originalFilename}`);
    console.log(`key: ${key}`);


    try {
      const command = new GetObjectCommand({
        Bucket: process.env.S3_BUCKET,
        Key: key,
      });

      const signedUrl = await getSignedUrl(s3, command, { expiresIn: 3600 });
      const fileResponse = await fetch(signedUrl);

      if (!fileResponse.ok) {
        throw new Error(`Error fetching from S3: ${fileResponse.statusText}`);
      }

      const fileBuffer = Buffer.from(await fileResponse.arrayBuffer());
      
      console.log(`DOC: ${doc.filename}`);

      await payload.update({
        collection: collectionSlug,
        id: doc.id,
        data: { alt: doc.alt || '' },
        overwriteExistingFiles: true,
        file: {
          data: fileBuffer,
          mimetype: doc.mimeType || 'application/octet-stream',
          size: fileBuffer.length,
          name: originalFilename
        },
      });

      console.log(`✅ Regenerated sizes for: ${originalFilename}`);
    } catch (error) {
      console.error(`❌ Failed to regenerate ${originalFilename}:`, error);
    }
  }

  console.log('🎉 Image regeneration complete!');
}

export async function down({}: MigrateDownArgs): Promise<void> {
  console.log('ℹ️ No rollback needed');
}

```

**Reply by nachomglz** [reply]
@camilo220 I don't know if you were able to solve this, but I just ran a script that updated each media (with no data updates, just use the update method for the media) and it re-generates the sizes and uploads the new images and everything, it ends up being easier than initially thought haha.

**Reply by nachomglz** [reply]
It's actually weird, because sometimes it does update the files in S3 and sometimes it doesn't...

**Reply by camilo220** [reply]
@nachomglz I wasn’t able to get the script above working, as you mentioned, it behaved differently each time I ran it. In the end, I used this script I found here:

https://www.reddit.com/r/PayloadCMS/comments/1it3ta6/payload_image_size_regeneration

Let me know how it works for you, and apologies for the delay in getting back to you.

**Comment by mvdve** [comment]
For anyone who is looking for an updated version of the script for local files (Payload 3.61.1) see below. Run with `npm run payload run xxx/regenerateMedia.ts`

```typescript
import config from '@payload-config';
import path from 'path';
import {getPayload} from 'payload';

async function regenerateMedia() {
    try {
        const payload = await getPayload({config});
        const media = await payload.find({
            collection: 'media',
            depth: 0,
            limit: 200,
        });

        if (!media.totalDocs) {
            console.log('No media found');
            return;
        }

        for (const mediaDoc of media.docs) {
            const staticDir = path.resolve(import.meta.filename, '../../../media');

            try {
                await payload.update({
                    collection: 'media',
                    id: mediaDoc.id,
                    data: mediaDoc,
                    overwriteExistingFiles: true,
                    filePath: `${staticDir}/${mediaDoc.filename}`,
                });

                console.log(`Media ${mediaDoc.id} (${mediaDoc.filename}) updated.`);
            } catch (err) {
                console.error(err);
            }
        }
    } catch (err) {
        console.error(err);
    }
}

await regenerateMedia();

```

**Comment by ChristopherNowlan** [comment]
Does anyone have a working version that updates images to S3
